### ISSUE TITLE
Help a novice with JS agent

### ISSUE THREAD

**Issue by fixbroke** [cantino/huginn#1112]
Hello! I'm just migrating my extensive Yahoo Pipes RSS filtering and item tweaking over to Huginn, and I'm very pleased so far.

In a couple of my item tweaks, I use a WebsiteAgent to do stuff like pull out an iframe from the article and append it to the item description. This works great, except that articles that have no iframe are not emitted by the WebsiteAgent as events. No problem there - I'll just duplicate the source events and send them and the WebsiteAgent events to a de-duplication agent. This works wonderfully, except that in many cases the iframe event from the WebsiteAgent gets filtered instead of the non-iframe event.

My Javascript experience is very limited, but I'd like to institute a delayed event emission with a Javascript agent so as to ensure the iframe events will arrive at the de-duplication agent first. Here's what I have as my current code:

```
Agent.receive = function() {

  setTimeout(unloaditems(),50000) //wait 50 seconds before continuing
}

function unloaditems() {
var events = Agent.incomingEvents();
  for(var i = 0; i < events.length; i++) {
      Agent.createEvent(events[i].payload);
    } 
}
```
However, this produces an error:
> E, [2015-10-27T16:09:56.523940 #20] ERROR -- : JavaScript error: setTimeout is not defined

How can I make this agent work as I intend? I'd also be interested to hear any alternate suggestions for my overall strategy of dealing with the missing events from the WebsiteAgent.

**Comment by cantino** [comment]
Hey @fixbroke, welcome!

I don't think that `setTimeout` is available in our JavaScript environment, provided by [therubyracer](https://github.com/cowboyd/therubyracer).

Maybe we can figure out a different solution.  When you say, "this works great, except that articles that have no iframe are not emitted by the WebsiteAgent as events," what do you mean?  What does the WebsiteAgent configuration look like for sites that do and do not have iframes?

**Comment by fixbroke** [comment]
Hey @cantino - thanks for starting this great project, and thanks for helping me with my problem.

Here's my Website Agent, the purpose of which is to extract the title image for articles in the RSS feed at http://www.motoiq.com/DesktopModules/DnnForge%20-%20NewsArticles/Rss.aspx?TabID=103&ModuleID=469
```{
  "expected_update_period_in_days": "4",
  "url": "{{link}}",
  "type": "xml",
  "mode": "merge",
  "extract": {
    "pic": {
      "css": ".articleEntry > p:first-child > img:first-child, .articleSummaryTitle:first-child > img:first-child",
      "value": "."
    }
  }
}```

Most linked articles don't have iframes, so no events are created for those. I would have thought merge mode would still pass the events through, but they're not emitted.

In a similar scenario (the aforementioned iframe extractor), I was fairly successful in achieving my goal by sending a duplicate, parallel event stream through a dummy event formatting agent without "propagate immediately" checked, and having the website agent and the dummy agent feed into into a de-duplication agent. However, this doesn't work as consistently in this case.

**Comment by cantino** [comment]
Did you mean to close this?

Where is an iframe involved in the above code?

**Comment by fixbroke** [comment]
Ha, no. Wrong button. 

The iframe extractor was a similar scenario, which works fairly well with the dummy agent/delayed propagation workaround. If I could get a javascript solution, it would apply to both.

**Comment by cantino** [comment]
So you're scraping one feed, and some items have iframes in them and others have images, and you want to combine it into an output that has both?

**Comment by fixbroke** [comment]
No, these are separate scenarios. Sorry for the confusion - I started writing about one scenario and then switched to the other, but missed a sentence where I mentioned iframes.

Both scenarios work the same way, in that all RSS item links are scraped by the website agent but only some articles have content pulled by the scrape. In the case of the code I posted, it's a lead article image (which is not in the original RSS feed) which I'm pulling from the linked article then appending to the description in a subsequent agent. Those articles that don't have lead images (or those that use a non-standard format for the image) don't get emitted by the website agent, but I'd still like to see those articles, so I have a parallel stream of events that are mostly duplicates.

You can see the scenario here:
https://huginnnnnnn.herokuapp.com/scenarios/7/export.json

**Comment by cantino** [comment]
It sounds like the basic issue is that you want the scraper to emit all events, some of which have images and some of which are empty, and right now you can't because if they don't have a matching image, they don't get emitted at all.  Is this right?  In that case, perhaps this is similar to https://github.com/cantino/huginn/issues/990?
